map-storage/src/Upload/UploadController.ts

```diff
@@ -0,0 +1,192 @@
+                // Iterate over the entries in the ZIP archive
+                for (const zipEntry of zipEntries) {
+                    // Store the file
+                    promises.push(
+                        this.fileSystem.writeFile(zipEntry, mapPath(path.join(directory, zipEntry.entryName), req), zip)
+                    );
+                }
+
+                await Promise.all(promises);
+
+                // Delete the uploaded ZIP file from the disk
+                fs.unlink(zipFile.path, (err) => {
+                    if (err) {
+                        console.error(`Error deleting file: ${err}`);
+                    }
+                });
+
+                res.send("File successfully uploaded.");
+            });
+
+            if (limiter.activeCount === 0 && limiter.pendingCount === 0) {
+                this.uploadLimiter.delete(virtualDirectory);
+            }
+        });
+    }
+
+    /**
+     * Let's filter out any file starting with "."
+     */
+    public filterFile(path: string): boolean {
+        const paths = path.split(/[\/\\]/);
+        for (const subPath of paths) {
+            if (subPath.startsWith(".")) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private getDownload() {
+        this.app.get("/download", passportAuthenticator, async (req, res) => {
+            const directoryRaw = req.query.directory;
+            const directory = z.string().optional().parse(directoryRaw) || "./";
+
+            if (directory.includes("..")) {
+                // Attempt to override filesystem. That' a hack!
+                res.status(400).send("Invalid directory");
+                return;
+            }
+
+            // Let's explicitly forbid 2 uploads from running at the same time using a p-limit of 1.
+            // Uploads will be serialized.
+            const virtualDirectory = mapPath(directory, req);
+
+            let archiveName: string | undefined;
+            if (directory) {
+                const baseName = path.basename(directory);
+                if (baseName !== "." && baseName !== "") {
+                    archiveName = baseName + ".zip";
+                }
+            }
+            if (archiveName === undefined) {
+                archiveName = "map.zip";
+            }
+
+            res.attachment(archiveName);
+
+            const archive = archiver("zip", {
+                zlib: { level: 9 }, // Sets the compression level.
+            });
+
+            // good practice to catch warnings (ie stat failures and other non-blocking errors)
+            archive.on("warning", function (err) {
+                if (err.code === "ENOENT") {
+                    // log warning
+                    console.warn("File not found: ", err);
+                } else {
+                    console.error("An warning occurred while Zipping file: ", err);
+                    //res.status(500).send('An error occurred');
```

Comment:
If there are warning, we never return status or body?

Reply:
I'm not sure we can do that. I'm "streaming" the ZIP file so I'm starting sending the answer BEFORE the full ZIP file is generated. We could have an issue in the middle of the ZIP. At this point, it is too late to change the HTTP status...